### PR TITLE
Redis cluster refactoring

### PR DESCRIPTION
**Pull request by hofrob**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #147 #143 #142 #136 #135 #134 #66

### CODE REVIEW

**Comment by hofrob** [comment]
Connections to redis are now pooled, to enable redirects without opening and closing connections during a request.

If redis is running in cluster mode, it is possible to supply a hashtag for `MGET` and `MSET` operations. If there are now hashtags, it falls back to `GET` and `SET`.

To test this locally I used a ready-to-use docker cluster container:

```
docker run --network=some_local_network --network-alias=redis-cluster grokzen/redis-cluster:redis_version
```

Put a PHP container with Yii in the same network (`some_local_network`), select a redis version (`redis_version`), and try [the example](https://github.com/yiisoft/yii2-redis/pull/178/files#diff-010266d5c06d24ba1968dd13647dc805R79).

**Comment by hofrob** [comment]
I should implement some tests for this change. Probably by using the aforementioned docker container. But to be honest, I have no idea where to start. If someone could point me in the right direction, I'd be happy to get started.

**Comment by hofrob** [comment]
@cebe 

This PR would fix the problem you mentioned here: https://github.com/yiisoft/yii2-redis/pull/147#discussion_r188220246

**Comment by cebe** [comment]
Thanks for your work on this. I am planning to use a redis cluster in the near future and will review this PR when I get to that.

**Comment by hofrob** [comment]
I can give you an update from what we're using now. Turns out this solution works, but somehow it was the slowest of all the solutions we tried (our production env runs on AWS EB with extensive use of redis ElastiCache). Our requests were just about 10-15% slower than before. But it was noticeable.

And yes, weirdly enough opening and closing the connections after redirects was faster. But I did not do any more research on that because....

We then tried the [redis pecl extension](https://github.com/phpredis/phpredis/). It supports cluster mode completely and we're now at about the same request duration we were before.

To be able to use the pecl extension with `\yii\caching\Cache` I just had to write a very simple wrapper (about 80 lines). The only thing left for us to completely switch to this extension is rewriting the yii2-queue. It's unfortunate that the method names are written differently.

* `hdel `vs `hDel`
* `lpush `vs `lPush`
* `zadd `vs `zAdd`

So I've not yet tackled this. I think it would be great if you could switch from `yii2-redis` to the pecl extension without having to change something in `yii2-queue`. If you or the yii devs in general are interested in implementing this, maybe I can help.

**Comment by samdark** [comment]
That is interesting, especially for 3.0.

**Comment by samdark** [comment]
@cebe did you have a chance to use Redis cluster?

**Comment by pacifier17** [comment]
@samdark do you think we can get this merged soon? We are looking to use cluster Redis for our use case. We did test out the fork and it works with cluster redis config.

File: src/Cache.php [hunk 1]
```diff
@@ -113,11 +129,25 @@ class Cache extends \yii\caching\Cache
...
      */
     private $_replica;
+    /**
+     * @var bool remember if redis is in cluster mode for the whole request
+     */
+    private $_isCluster;
+    /**
+     * @var string check if hash tags were supplied for a MGET/MSET operation
+     */
+    private $hashTagAvailable = false;
```
Comment (samdark) [inline src/Cache.php]:
```suggestion
    private $_hashTagAvailable = false;
```

File: src/Cache.php [hunk 2]
```diff
@@ -159,16 +189,37 @@ protected function getValue($key)
      */
     protected function getValues($keys)
     {
+        if ($this->isCluster && !$this->hashTagAvailable) {
```
Comment (samdark) [inline src/Cache.php]:
```suggestion
        if ($this->isCluster && !$this->_hashTagAvailable) {
```

File: src/Cache.php [hunk 3]
```diff
@@ -159,16 +189,37 @@ protected function getValue($key)
...
     }
 
+    public function buildKey($key)
+    {
+        if (
+            is_string($key)
+            && $this->isCluster
+            && preg_match('/^(.*)(\{.+\})(.*)$/', $key, $matches) === 1) {
+
+            $this->hashTagAvailable = true;
```
Comment (samdark) [inline src/Cache.php]:
```suggestion
            $this->_hashTagAvailable = true;
```

File: src/Cache.php [hunk 4]
```diff
@@ -188,6 +239,10 @@ protected function setValue($key, $value, $expire)
      */
     protected function setValues($data, $expire)
     {
+        if ($this->isCluster && !$this->hashTagAvailable) {
```
Comment (samdark) [inline src/Cache.php]:
```suggestion
        if ($this->isCluster && !$this->_hashTagAvailable) {
```

File: src/Cache.php [hunk 5]
```diff
@@ -215,9 +270,34 @@ protected function setValues($data, $expire)
             }
         }
 
+        $this->hashTagAvailable = false;
```
Comment (samdark) [inline src/Cache.php]:
```suggestion
        $this->_hashTagAvailable = false;
```

File: src/Cache.php [hunk 6]
```diff
@@ -215,9 +270,34 @@ protected function setValues($data, $expire)
...
+            $this->_isCluster = false;
+            try {
+                /**
+                 * if redis is running without cluster support, this command results in:
+                 * `ERR This instance has cluster support disabled`
+                 * and [[Connection::executeCommand]] throws an exception
+                 */
+                $this->redis->executeCommand('CLUSTER INFO');
+                $this->_isCluster = true;
+            } catch (Exception $exception) {
```
Comment (samdark) [inline src/Cache.php]: If the intent is to ignore every exception possible then it should be added as comment inside `catch`

File: src/Connection.php [hunk 7]
```diff
@@ -242,6 +246,10 @@ class Connection extends Component
      * If [[unixSocket]] is specified, hostname and [[port]] will be ignored.
      */
     public $hostname = 'localhost';
+    /**
+     * @var string if the query gets redirected, use this as the temporary new hostname
```
Comment (samdark) [inline src/Connection.php]: `@since 2.0.11` should be added

File: src/Cache.php [hunk 8]
```diff
@@ -113,11 +129,25 @@ class Cache extends \yii\caching\Cache
      * @see $enableReplicas
      */
     public $replicas = [];
+    /**
+     * @var bool|null force cluster mode, don't check on every request. If this is null, cluster mode will be checked
+     *                once per request whenever the cache is accessed. To disable the check, set to true if cluster
+     *                mode should be enabled, or false if it should be disabled.
```
Comment (samdark) [inline src/Cache.php]: `@since 2.0.11` should be added

File: src/Cache.php [hunk 9]
```diff
@@ -215,9 +270,34 @@ protected function setValues($data, $expire)
             }
         }
 
+        $this->hashTagAvailable = false;
+
         return $failedKeys;
     }
 
+    public function getIsCluster()
```
Comment (samdark) [inline src/Cache.php]: `@since 2.0.11` should be added along with phpdoc

File: src/Connection.php [hunk 10]
```diff
@@ -520,13 +527,27 @@ public function __sleep()
         return array_keys(get_object_vars($this));
     }
 
+    public function getConnectionString()
```
Comment (samdark) [inline src/Connection.php]: `@since 2.0.11` should be added along with phpdoc

File: src/Connection.php [hunk 11]
```diff
@@ -520,13 +527,27 @@ public function __sleep()
...
+    {
+        if ($this->unixSocket) {
+            return 'unix://' . $this->unixSocket;
+        }
+
+        return 'tcp://' . ($this->redirectConnectionString ?: "$this->hostname:$this->port");
+    }
+
+    public function getSocket()
```
Comment (samdark) [inline src/Connection.php]: `@since 2.0.11` should be added along with phpdoc

**Comment by samdark** [comment]
Overall it looks OK. Some minor adjustments and changelog line are needed.

**Comment by samdark** [comment]
Unit tests would be awesome as well.

**Comment by samdark** [comment]
@hofrob would you please apply suggestions?

File: src/Cache.php (hunk 6 above)
Comment (hofrob) [reply src/Cache.php]: The intention is to catch this exception: https://github.com/yiisoft/yii2-redis/blob/a456513924fa448ac84bdb2d6c200f194509bce2/src/Connection.php#L746

I didn't want not change the way `parseResponse` handles actual errors. The comment in the try explains what I'm doing.

**Comment by hofrob** [comment]
> changelog line are needed.

Since this is a really old branch, is there a preferred way how to update it? I'd just rebase it (and then add the changelog).

**Comment by samdark** [comment]
Either merging master into it or rebasing.

**Comment by hofrob** [comment]
> Unit tests would be awesome as well.

After a quick look at the sources I'm not really sure how to configure/start/use a redis cluster for tests.

**Comment by samdark** [comment]
Merged. Thank you very much!

**Comment by pacifier17** [comment]
Thanks a lot @samdark and @hofrob for merging these changes in so quickly. I really appreciate it.

**Comment by samdark** [comment]
Not **so quickly**. Looks how much time it took me to get to it and review/merge it :)  @hofrob sorry for that btw.

**Pull request merged**